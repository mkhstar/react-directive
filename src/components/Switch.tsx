import React, { ReactNode } from "react";
import { SwitchProps } from "../types/switch";

export const Switch: React.FC<SwitchProps> = ({ children }) => {
  let caseToRender: ReactNode | null = null;

  React.Children.forEach(children, (child) => {
    if (caseToRender) return;

    const { props } = child;
    if (
      "when" in props &&
      (typeof props.when === "function" ? props.when() : props.when)
    ) {
      caseToRender = child;
      return;
    }
    if (!("when" in props)) caseToRender = child;
  });

  return caseToRender;
};
