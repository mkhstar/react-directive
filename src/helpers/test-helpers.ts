import { ReactElement } from "react";
// eslint-disable-next-line import/no-extraneous-dependencies
import renderer from "react-test-renderer";

export const snapshotTest = (
  render: ReactElement<any, string>,
  title?: string
) => {
  const tree = renderer.create(render).toJSON();
  expect(tree).toMatchSnapshot(title || "");
};

export const truthyValues = ["value", true, 1, Infinity, -1, -Infinity, {}, []];
export const falsyValues = ["", false, 0, null];
export const truthyFalsyMap = truthyValues
  .map((value) =>
    falsyValues.map((falsyValue) => ({ truthy: value, falsy: falsyValue }))
  )
  .flat();
