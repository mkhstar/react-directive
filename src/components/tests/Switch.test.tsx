import React from "react";
import { render, cleanup } from "@testing-library/react";
import { snapshotTest, truthyFalsyMap } from "../../helpers/test-helpers";
import { Switch, Case, Default } from "../Switch";

afterEach(cleanup);

describe("Switch, Case and Default", () => {
  describe("<Switch />", () => {
    it("renders the first Case whose when prop returns truthy value", () => {
      truthyFalsyMap.forEach((value) => {
        const { getByText, queryByText, unmount } = render(
          <Switch>
            <Case when={value.falsy}>
              <div>Case 1</div>
            </Case>
            <Case when={value.truthy}>
              <div>Case 2</div>
            </Case>
            <Case when={value.truthy}>
              <div>Case 3</div>
            </Case>
            <Default>
              <div>Default</div>
            </Default>
          </Switch>
        );
        expect(queryByText("Case 1")).toBeNull();
        expect(getByText("Case 2")).toBeInTheDocument();
        expect(queryByText("Case 3")).toBeNull();
        expect(queryByText("Default")).toBeNull();
        unmount();
      });
    });

    it("renders the first Case whose when prop function returns true", () => {
      truthyFalsyMap.forEach((value) => {
        const { getByText, queryByText, unmount } = render(
          <Switch>
            <Case when={() => value.falsy}>
              <div>Case 1</div>
            </Case>
            <Case when={() => value.truthy}>
              <div>Case 2</div>
            </Case>
            <Case when={value.truthy}>
              <div>Case 3</div>
            </Case>
            <Default>
              <div>Default</div>
            </Default>
          </Switch>
        );
        expect(queryByText("Case 1")).toBeNull();
        expect(getByText("Case 2")).toBeInTheDocument();
        expect(queryByText("Case 3")).toBeNull();
        expect(queryByText("Default")).toBeNull();
        unmount();
      });
    });

    it("renders the Default component if no Case's when prop returns true", () => {
      const { getByText, queryByText, unmount } = render(
        <Switch>
          <Case when={false}>
            <div>Case 1</div>
          </Case>
          <Case when={false}>
            <div>Case 2</div>
          </Case>
          <Default>
            <div>Default</div>
          </Default>
        </Switch>
      );
      expect(queryByText("Case 1")).toBeNull();
      expect(queryByText("Case 2")).toBeNull();
      expect(getByText("Default")).toBeInTheDocument();
      unmount();
    });
  });

  describe("<Case />", () => {
    it("renders children if when prop returns true", () => {
      const { getByText, unmount } = render(
        <Case when>
          <div>Case</div>
        </Case>
      );
      expect(getByText("Case")).toBeInTheDocument();
      unmount();
    });

    it("renders children if when prop function returns true", () => {
      const { getByText, unmount } = render(
        <Case when={() => true}>
          <div>Case</div>
        </Case>
      );
      expect(getByText("Case")).toBeInTheDocument();
      unmount();
    });

    it("does not render children if when prop returns false", () => {
      const { queryByText, unmount } = render(
        <Case when={false}>
          <div>Case</div>
        </Case>
      );
      expect(queryByText("Case")).toBeNull();
      unmount();
    });
  });

  describe("<Default />", () => {
    it("Should render whatever child it receives", () => {
      const { getByText, unmount } = render(
        <Default>
          <div>Default</div>
        </Default>
      );
      expect(getByText("Default")).toBeInTheDocument();
      unmount();
    });
  });
});

describe("<Switch /> snapshots", () => {
  it("matches the snapshot with the matching case", () => {
    snapshotTest(
      <Switch>
        <Case when>
          <div>Case 1</div>
        </Case>
        <Case when={false}>
          <div>Case 2</div>
        </Case>
      </Switch>
    );
    snapshotTest(
      <Switch>
        <Case when={false}>
          <div>Case 1</div>
        </Case>
        <Case when>
          <div>Case 2</div>
        </Case>
        <Case when>
          <div>Case 3</div>
        </Case>
        <Default>
          <div>Default</div>
        </Default>
      </Switch>
    );
  });

  it("matches the snapshot with the default case", () => {
    snapshotTest(
      <Switch>
        <Case when={false}>
          <div>Case 1</div>
        </Case>
        <Default>
          <div>Default case</div>
        </Default>
      </Switch>
    );
    snapshotTest(
      <Switch>
        <Case when={false}>
          <div>Case 1</div>
        </Case>
      </Switch>,
      "No default case"
    );
  });
});
