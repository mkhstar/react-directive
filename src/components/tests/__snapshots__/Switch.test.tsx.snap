// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> snapshots matches the snapshot with the default case 1`] = `
<div>
  Default case
</div>
`;

exports[`<Switch /> snapshots matches the snapshot with the default case: No default case 1`] = `null`;

exports[`<Switch /> snapshots matches the snapshot with the matching case 1`] = `
<div>
  Case 1
</div>
`;

exports[`<Switch /> snapshots matches the snapshot with the matching case 2`] = `
<div>
  Case 2
</div>
`;
